New model of gene-sample expression value and management command to load data sheet into database

dongbohu opened the pull request:
Although this PR includes 4 files, you can skip these two:
* migrations/0006_expressionvalue.py (generated by database migration)
* data/all-pseudomonas-gene-normalized.pcl (datasheet provided by Jie)

`models.py` includes a new class to model the gene-sample relationship.
`import_gene_sample_expr.py` is the management command to read the data sheet into backend database.

dongbohu commented:
I also updated database schema at: 
https://sketchboard.me/rzMAA9ecMFEh#
The box whose border is red is the class that I added in `models.py`.

dongbohu commented:
The new issues in Codeclimate are probably due to the mixture of if/else and try/except.

dongbohu commented on adage/analyze/management/commands/import_gene_sample_expr.py:
```diff
@@ -0,0 +1,140 @@
+    using "ml_data_source" field. If a token does not match any sample's
+    ml_data_source, put None into "samples".)
+
+    An exception will be raised if any of the following errors is detected:
```
Believe it or not, I originally used "are" but changed to "is" later. There are some interesting discussions on this topic:
https://www.englishforums.com/English/AnyIsAre/hxhhn/post.htm
Most people seem to think that both `are` and `is` are okay.

kurtwheeler replied:
English is weird.

mhuyck replied:
The plural form is correct.

For such grammatical "edge cases," I usually turn to *The Elements of Style*, by William Strunk, Jr., and E.B. White, more commonly known as "Strunk & White." The relevant rule is described in Strunk & White in Section I, part 9: "The number of the subject determines the number of the verb." In this case, the part that's tricky is that the subject of the sentence is "any of."

The example given in the book is:

> A common blunder is the use of a singular verb form in a relative clause following "one of ..." or a similar expression when the relative is the subject.

and then two examples are given showing the wrong wording in a column on the left and the corrected wording in a column on the right. Rather than type them all out, the corrected first example says

> One of the ablest men who have attacked this problem

But we are not writing a thesis here, so I'm good with it either way. Just thought you'd like to know about Strunk & White. It is considered an authoritative source and a worthwhile companion for any writer of English.

rzelayafavila commented on adage/analyze/models.py:
```diff
@@ -218,3 +218,17 @@ def __unicode__(self):
+    """
+    sample = models.ForeignKey(Sample, on_delete=models.PROTECT)
+    gene = models.ForeignKey(Gene, on_delete=models.PROTECT)
+    value = models.FloatField()
```
Nice, neat model, and good, descriptive docstring too!

tj8901nm replied:
I don't think gene expression values have units. They are relative measurements. To make it a little more precise, you can say "RMA processed and zero-one normalized gene expression values".

mhuyck replied:
Right, I guess if it's normalized then the units are taken out. Your description of the pre-processing is exactly what I was looking for.

dongbohu replied:
Thank you both for the comments. I will change it.

rzelayafavila commented on adage/analyze/models.py:
```diff
@@ -218,3 +218,17 @@ def __unicode__(self):
+    value = models.FloatField()
+
+    def __unicode__(self):
+        return "Sample %s and Gene %s with expression value %f" % (
```
Hmm, I think I would phrase this more like "Expression value %f for Sample %s and Gene %s", so that the emphasis is on the fact that this is an Expression value. What do you think?

dongbohu replied:
Sure. I like that.

rzelayafavila commented on adage/analyze/management/commands/import_gene_sample_expr.py:
```diff
@@ -0,0 +1,140 @@
+            records.append(
+                ExpressionValue(sample=sample, gene=gene, value=float_val))
+        col_num += 1
+    ExpressionValue.objects.bulk_create(records)  # Create records in bulk.
```
:+1: for ``bulk_create``

rzelayafavila commented on adage/analyze/management/commands/import_gene_sample_expr.py:
```diff
@@ -0,0 +1,140 @@
+
+"""
+This management command reads an input file of gene-sample expression values
+and loads the valid data into database.  It should be invoked like this:
```
Quick minor thing: I think this sentence would flow better if it was either "... and loads the valid data into **the** database" or "... and loads valid data into database." I kind of prefer this first one, though.

dongbohu replied:
Sure. I will change it to the first one.

rzelayafavila commented on adage/analyze/management/commands/import_gene_sample_expr.py:
```diff
@@ -0,0 +1,140 @@
+                        " is blank" % line_num)
+
+    try:
+        gene = Gene.objects.get(systematic_name=gene_name)
```
Hmmm... is this command agnostic to the organism/species? What if there are two genes with the same systematic_name from different species?

dongbohu replied:
Good question. I never thought of this. Is it possible that two different organisms may have different genes whose systematic names are identical?

mhuyck replied:
Yes, unfortunately it is. I think we may need to specify an organism on the command line. Biology is even more complicated than English. 😁

dongbohu replied:
Thanks. I will add an argument as input organism.

rzelayafavila commented on adage/analyze/management/commands/import_gene_sample_expr.py:
```diff
@@ -0,0 +1,140 @@
+    with transaction.atomic():
+        samples = []
+        for line_index, line in enumerate(file_handle):
+            tokens = line.rstrip('\r\n').split('\t')
```
Why are we specifying ``'\r\n'`` in ``rstrip()`` here? Is it so that rstrip will not remove single spaces and tabs?

dongbohu replied:
I think it is also okay to simply use `rstrip()`, but I want to keep the trailing spaces and tabs (which will be detected later) so that the error will be "blank gene names" instead of "number of columns is not xxx". Does this make sense?

rzelayafavila replied:
Yep, that's what I imagined, sounds good.

mhuyck commented on adage/analyze/management/commands/import_gene_sample_expr.py:
```diff
@@ -0,0 +1,140 @@
+
+    if len(tokens) != len(samples) + 1:
+        raise Exception("Input file line #%d: Number of columns is not %d" %
+                        (line_num, len(samples) + 1))
```
Thank you for reporting line numbers! I know you always do, but want to make sure you know that the effort is greatly appreciated. This type of error always seems to crop up when it's most helpful to have a quick solution.

dongbohu replied:
Thank you for the recommendation of `The Elements of Style` book, it is definitely something good to know. But I am not 100% convinced that the rule in Section I, part 9 applies to the current scenario, because in "any of the following errors is/are detected", the subject is "any", which could refer to either singular or plural, while in:
```One of the ablest men who have attacked this problem```
"who" clearly refers to "men", which is plural. 
As you said, this is just a comment line and not that important, so I will change it into plural.

dongbohu commented:
I think I have addressed all concerns from the reviewers. I also added a `unique_together` restriction in the new model (and another migration file, which you can ignore). Please  take a look again and let me know what you think. Thanks.

mhuyck commented on adage/analyze/management/commands/import_gene_sample_expr.py:
```diff
@@ -2,19 +2,35 @@
+to add it.
+(2) If a data source (on the first row of input file) or gene name (on the
+first column of input file) is not found in the database, a warning message
+will be generated and the corresponding column or row will be skipped.
```
Thank you for the excellent documentation!

rzelayafavila replied:
:+1:

mhuyck commented on adage/analyze/management/commands/import_gene_sample_expr.py:
```diff
@@ -40,12 +58,21 @@ def handle(self, **options):
+        raise Exception("Input organism_tax_id is not found in the database. "
+                        "Please use the management command "
+                        "'organism_create_or_update.py' in djago-organism "
+                        "package to create this organism.")
```
A user-friendly error message with concrete steps to solve the problem. Awesome! :100:

rzelayafavila replied:
:100:

dongbohu replied:
Just noticed the typo: `djago-organism` should be `django-organisms`. I will fix it.

dongbohu replied:
Found a similar typo in `add_ml_model.py`, fixed it too and included it in this PR.

mhuyck commented on adage/analyze/management/commands/import_gene_sample_expr.py:
```diff
@@ -67,7 +94,7 @@ def read_header(tokens, samples):
     ml_data_source, put None into "samples".)
 
-    An exception will be raised if any of the following errors is detected:
+    An exception will be raised if any of the following errors are detected:
```
👍

mhuyck commented on adage/analyze/management/commands/import_gene_sample_expr.py:
```diff
@@ -109,17 +136,21 @@ def import_data_line(line_num, tokens, samples):
-            "Input file line #%d: gene systematic_name (column #1) is not "
-            "found in the database: %s", line_num, gene_name)
+            "Input file line #%d: gene name %s (column #1) not found in "
+            "database", line_num, gene_name)
```
Maybe I'm missing something, but don't we need this to say:

```python
...database" % (line_num, gene_name))
```

(as in lines 145-147 above) and I think line 151 also needs to be `logger.warning(`, right?

dongbohu replied:
I checked python doc and seems that both formats of the arguments will work:
https://docs.python.org/2/library/logging.html#logging.Logger.debug
The arguments of `warning()` method is same as those in `debug()` method.
(I think I read an online article some time ago which argues that the format that I am using has better performance, but I couldn't find the link now).

You are definitely right about the method name `warning`. `warn()` is a method for `warnings` library.
https://docs.python.org/2/library/warnings.html#warnings.warn

Two other management commands `import_data.py` and `import_activity.py` are also using `warn()` instead of `warning()`, I will file another PR after this one to fix them.

dongbohu replied:
Wait a minute: I just remembered that when I used `logger.warn()`, the warning messages are correctly printed out on the screen during importing process. Let me read the python doc more carefully.

dongbohu replied:
`warn()` seems to be a "shortcut" of  `warning()` in `logger` class, but I don't understand why python online doc doesn't mention the former.

dongbohu replied:
I am going to change `warn` to `warning` in this file to avoid confusion in the future.

mhuyck replied:
Sounds good. If it's not documented it might go away in the future without `warn`ing. (Heh.)

mhuyck commented:
By the way, the `unique_together` constraint is a very good idea. Thanks for catching that, too!

mhuyck commented:
Looks great. Thank you @dongbohu!
